pkg/scheduler/controller/scheduler_control.go

```diff
@@ -0,0 +1,375 @@
+			seedRegion := seed.Spec.Cloud.Region
+
+			for currentMaxMatchingCharacters < len(shootRegion) {
+				if strings.HasPrefix(seedRegion, shootRegion[:currentMaxMatchingCharacters+1]) {
+					candidates = []*gardenv1beta1.Seed{}
+					currentMaxMatchingCharacters++
+					continue
+				} else if strings.HasPrefix(seedRegion, shootRegion[:currentMaxMatchingCharacters]) {
+					candidates = append(candidates, seed)
+				}
+				break
+			}
+		}
+	}
+	return candidates
+}
+
+func generateSeedUsageMap(shootList []*gardenv1beta1.Shoot) map[string]int {
+	m := map[string]int{}
+
+	for _, shoot := range shootList {
+		if seed := shoot.Spec.Cloud.Seed; seed != nil {
+			m[*seed]++
+		}
+	}
+
+	return m
+}
+
+func validateDisjointedNetworks(seed *gardenv1beta1.Seed, shoot *gardenv1beta1.Shoot) (bool, field.ErrorList) {
+	// error cannot occur due to our static validation
+	k8sNetworks, _ := gardenhelper.GetK8SNetworks(shoot)
+	allErrs := schedulerutils.ValidateNetworkDisjointedness(seed.Spec.Networks, k8sNetworks, field.NewPath(""))
+	return len(allErrs) == 0, allErrs
+}
+
+func verifySeedAvailability(seed *gardenv1beta1.Seed) bool {
+	if cond := gardencorev1alpha1helper.GetCondition(seed.Status.Conditions, gardenv1beta1.SeedAvailable); cond != nil {
+		return cond.Status == gardencorev1alpha1.ConditionTrue
+	}
+	return false
+}
+
+// UpdateShootToBeScheduledOntoSeed (ctx context.Context, cl client.Client, client kubernetes.Interface, shoot *gardenv1beta1.Shoot, seed *gardenv1beta1.Seed) error {
+func UpdateShootToBeScheduledOntoSeed(ctx context.Context, shoot *gardenv1beta1.Shoot, seed *gardenv1beta1.Seed, createShootBinding createShootBinding) error {
+	// API Server does not allow to update seed directly. We need to use a ShootBinding Object.
+	binding := &gardenv1beta1.ShootBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      shoot.Name,
+			Namespace: shoot.Namespace,
+		},
+		Target: seed,
+	}
+	return createShootBinding(ctx, binding)
+}
+
+func (c *defaultControl) updateShootStatusToUnschedulable(ctx context.Context, shoot *gardenv1beta1.Shoot, err error, condition gardencorev1alpha1.Condition) error {
+	// the shoot can't be scheduled to a seed right now, for example due to a lack of matching seeds in the desired region
+	message := fmt.Sprintf("Failed to schedule shoot '%s' : %+v", shoot.Name, err)
+	return c.updateShootStatusWithCondition(ctx, shoot, message, condition)
+}
+
+func (c *defaultControl) updateShootStatusWithCondition(ctx context.Context, shoot *gardenv1beta1.Shoot, message string, condition gardencorev1alpha1.Condition) error {
+	conditionUnschedulable := gardencorev1alpha1helper.UpdatedCondition(condition, gardencorev1alpha1.ConditionFalse, "Unschedulable", message)
+	return c.updateShootStatus(ctx, shoot, conditionUnschedulable)
+}
+
+func (c *defaultControl) updateShootStatus(ctx context.Context, shoot *gardenv1beta1.Shoot, conditions ...gardencorev1alpha1.Condition) error {
```

Comment:
The scheduler should not update the shoot status - just write `.spec.cloud.seed` and send events